system/jlib/jstats.cpp

```diff
@@ -1994,12 +2098,95 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
             cur.visit(visitor);
     }
 
+    virtual bool refreshAggregates(const StatisticsMapping & mapping) override
+    {
+        if (isDirty==false)
+            return false;
+        CRuntimeStatisticCollection totals(mapping);
+        Owned<IBitSet> totalUpdated = createBitSet(mapping.numStatistics());
+        return refreshAggregates(totals, *totalUpdated);
+    }
+
+    virtual bool refreshAggregates(CRuntimeStatisticCollection & totals, IBitSet & isTotalUpdated) override
+    {
+        const StatisticsMapping & mapping = totals.queryMapping();
+        bool updated = false;
+        // if this scope is not dirty, the aggregates are accurate at this level so return totals (no need to descend)
+        // Also if at sg scope, do not descend as aggregates does not need to be generated from below sg level
+        // (if aggregates should be calculated from descendants of sg scope, then remove the second test)
+        if (isDirty==false || id.queryScopeType()==SSTsubgraph)
+        {
+            // return the aggregates at this scope level
+            ForEachItemIn(i, stats)
+            {
+                Statistic & stat = stats.element(i);
+                StatisticKind kind = stat.queryKind();
+                if (kind != (StatisticKind)(kind & StKindMask))
+                    continue; // ignore variants->not supported by CRuntimeStatisticCollection
+                unsigned index = mapping.getIndex(kind);
+                if (index!=mapping.numStatistics())
+                {
+                    totals.mergeStatistic(kind, stat.queryValue());
+                    isTotalUpdated.set(index);
+                    updated = true;
+                }
+            }
+        }
+        else
+        {
+            // descend down to lower level to obtain totals required for aggregates and then aggregate
+            CRuntimeStatisticCollection childTotals(mapping);
+            Owned<IBitSet> childTotalUpdated = createBitSet(mapping.numStatistics());
```

Comment:
This logic does not seem correct.  total will only updated with any values that have changed - but in the not-dirty case all values are updated.

Reply:
Agree, I don't follow the logic (related to previous comment here: https://github.com/hpcc-systems/HPCC-Platform/pull/17786#discussion_r1377617622).

The true condition will be followed unconditionally for all subgraphs (regardless of whether isDirty = true or false), which looks like it means everything above every subgraph will also be updated.
Maybe

Reply:
I think it achieves the goal, but am still finding how it is doing it a bit confusing.
What I'd expect I think, is for each level to do:
+ if at level and dirty, do nothing
+ If at level and is subgraph, do nothing.
+ If dirty, recurse down each scope, and then merge the child values into new totals at current level.
+ When finished recursing down each child scope and merging into new totals, update what has changed into current level stats, and push totals (via callback if necessasry).


I had a stab at what that would look like, and I also think it should be done without adding the two quite implementation orientated 'refreshAggregates' methods to IStatisticCollection : 

```
bool refreshAggregates(IStatisticCollection &stats, const StatisticsMapping & mapping, AggregateUpdatedCallBackFunc & fWhenAggregateUpdated)
{
    if (stats.queryScopeType() == SSTsubgraph) // don't descend any further
        return false;
    if (!stats.isDirty())
        return false;
    bool anyChanges = false;
    CRuntimeStatisticCollection totals(mapping);
    Owned<IBitSet> updated = createBitSet(mapping.numStatistics());
    Owned<IStatisticCollectionIterator> childIter = &stats.getScopes(nullptr, false);
    ForEach(*childIter)
    {
        IStatisticCollection &child = childIter->query();
        refreshAggregates(child, mapping, fWhenAggregateUpdated);

        // merge in child level stats into current level, and mark which aggregate kinds have updated
        unsigned numChildStats = child.getNumStatistics();
        for (unsigned s=0; s<numChildStats; s++)
        {
            StatisticKind kind;
            unsigned __int64 value;
            child.getStatistic(kind, value, s);
            if (kind != (StatisticKind)(kind & StKindMask))
                continue; // ignore variants->not supported by CRuntimeStatisticCollection
            unsigned index = mapping.getIndex(kind);
            if (index != mapping.numStatistics())
            {
                totals.mergeStatistic(kind, value);
                updated->set(index); // NB: may already have been set whilst merging in another child, but that's okay.
            }
        }
    }

    // any updates? Set new totals into current level, and call callback
    unsigned nextUpdateIndex = 0;
    while (true)
    {
        nextUpdateIndex = updated->scan(nextUpdateIndex, true);
        if (NotFound == nextUpdateIndex)
            break;
        
        StatisticKind kind = mapping.getKind(nextUpdateIndex);
        unsigned __int64 value = totals.queryStatisticByIndex(nextUpdateIndex).get();
        if (stats.updateStatistic(kind, value, StatsMergeReplace))
        {
            if (value || includeStatisticIfZero(kind))
            {
                StringBuffer s;
                fWhenAggregateUpdated(stats.getFullScope(s).str(), stats.queryScopeType(), kind, value);
                anyChanges = true;
            }
        }
    }
    stats.clearDirty();
    return anyChanges;
}
```